File: LY/LYLeetCode/SpiralMatrixII.cpp
/*
 Input: 3
 Output:
 [
 [ 1, 2, 3 ],
 [ 8, 9, 4 ],
 [ 7, 6, 5 ]
 ]
 */

#include <iostream>

int** generateMatrix(int n) {
    int **matrix=(int **)malloc(sizeof(int*)*n);
    for (int i=0; i<n; i++) {
        matrix[i]=(int *)malloc(sizeof(int)*n);
    }
    int curretNum=1;
    //p记录本次要走的步数
    int p=n-1;
    for (int i=0; i<n/2; i++) {
        //int  count = p-1;
        for (int col=i; col<i+p; col++) {
            matrix[i][col]=curretNum++;
        }
        for (int row=i; row<i+p; row++) {
            matrix[row][i+p]=curretNum++;
        }
        for (int col=i+p; col>i; col--) {
            matrix[i+p][col]=curretNum++;
        }
        for (int row=i+p; row>i; row--) {
            matrix[row][i]=curretNum++;
        }
        p-=2;
    }
    if (n%2) {
        matrix[n/2][n/2]=curretNum;
    }
    return matrix;
}

int main(int argc, const char * argv[]) {
    int** result=generateMatrix(5);
    for (int i=0; i<5; i++) {
        for (int j=0; j<5; j++) {
            printf("%d ",result[i][j]);
        }
        printf("\n");
    }
}

File: CJN/CJNLeetCode/SpiralMatrixII.java
class Solution {
    public int[][] generateMatrix(int n) {
         int[][] matrix= new int[n][n];
        int count=1;
        int i=0;
        int j=0;
        matrix[i][j]=count++;
        while(count<=Math.pow(n,2)){
            while(j<matrix[i].length-1&&matrix[i][j+1]==0)
                matrix[i][++j]=count++;
            while(i<matrix.length-1 && matrix[i+1][j]==0)
                matrix[++i][j]=count++;
            while(j>0 && matrix[i][j-1]==0)
                matrix[i][--j]=count++;
            while(i>0&&matrix[i-1][j]==0)
                matrix[--i][j]=count++;
        }
        return matrix;
    }
}
